### PR TITLE
Feature/lifetime of job application and escalation

### PR DESCRIPTION
**Pull request by kanuj365**
Targets #2686

### Description
-  Created a scheduler to send email notifications.
-  Fetched the count of total applications with the status new and in progress was received 15 days before the current time.
-  Created a config variable, for the outdated time period.
-   Sent an email on created email template to hr.

### Checklist:

- [x] I have performed a self-review of my own code.

### CODE REVIEW

**Comment by kanuj365** [comment]
image

File: resources/views/emails/send-application-lifecycle.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<div>
+<h4> Hi HR team, </h4>
+
+<p>We have a total of <span>{{ $no_of_application }}</span> applications that are outdated and no action is being taken on them.</p>
```
Comment (P4NK4J) [inline resources/views/emails/send-application-lifecycle.php:4]: let's use camel casing for variables here

File: app/Console/Kernel.php [hunk 2]
```diff
@@ -29,6 +30,8 @@ class Kernel extends ConsoleKernel
         FixedBudgetProject::class,
         SendDailyEffortSummaryForProjectsOnGoogleChat::class,
         RemindProjectMembersToUpdateEffortOnGoogleChat::class,
+        Lifecycle_of_mail::class,
```
Comment (P4NK4J) [inline app/Console/Kernel.php:33]: Naming convention is not correct here as well

File: Modules/HR/Emails/Lifecycle_of_mail.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
+<?php
+
+namespace Modules\HR\Emails;
+
+use Illuminate\Console\Command;
+use Illuminate\Support\Facades\Mail;
+use Modules\HR\Entities\Application;
+
+class Lifecycle_of_mail extends Command
```
Comment (P4NK4J) [inline Modules/HR/Emails/Lifecycle_of_mail.php:9]: Please see how laravel classes are written, what are the naming conventions, and update accordingly. Also, check the file name casing

File: Modules/HR/Emails/Lifecycle_of_mail.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        $application_dates = Application::whereIn('status', ['new', 'in_progress'])->pluck('created_at');
+        $exceed_date_application_no = 0;
+        foreach ($application_dates as $date) {
+            $difference_days = $date->diffInDays(now());
+            if ($difference_days > config('hr.time-period.outdated')) {
+                $exceed_date_application_no += 1;
+            }
+        }
+
+        return Mail::send('emails.send-application-lifecycle', ['no_of_application' => $exceed_date_application_no], function ($messge) use ($email) {
```
Comment (P4NK4J) [inline Modules/HR/Emails/Lifecycle_of_mail.php:52]: Let's use queues to send the email here

File: Modules/HR/Emails/Lifecycle_of_mail.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    /**
+     * Execute the console command.
+     *
+     * @return int
+     */
+    public function handle()
+    {
+        $email = 'hr@coloredcow.com';
```
Comment (P4NK4J) [inline Modules/HR/Emails/Lifecycle_of_mail.php:42]: Let's fetch this email address from ENV file, you can refer to other places where its sued.

File: Modules/HR/Emails/Lifecycle_of_mail.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        $application_dates = Application::whereIn('status', ['new', 'in_progress'])->pluck('created_at');
+        $exceed_date_application_no = 0;
```
Comment (P4NK4J) [inline Modules/HR/Emails/Lifecycle_of_mail.php:44]: Correct naming convention in variables

Comment (P4NK4J) [inline Modules/HR/Emails/Lifecycle_of_mail.php:44]: what's the use of this variable ? Its name is very confusing, can we rename it to something meaningful ?

**Review by P4NK4J — CHANGES_REQUESTED** [review]
Left some comments, please check

File: Modules/HR/Emails/Lifecycle_of_mail.php (hunk 6 above)
Comment (kanuj365) [reply Modules/HR/Emails/Lifecycle_of_mail.php]: that variable take only those value whose date exceeded the time limit of application submission date.

File: Modules/HR/Config/config.php [hunk 7]
```diff
@@ -15,6 +15,12 @@
         'cut-off-score-resume-screening' => 6,
         'cut-off-score-telephonic-interview' => 5,
     ],
+    'time-period'=>[
+        'outdated'=>15
+    ],
+    'applications-life-cycle' => [
+    'email' => 'hr@coloredcow.com'
+    ],
```
Comment (P4NK4J) [inline Modules/HR/Config/config.php:23]: This is supposed to be fetched from ENV, just refer to other places how we are using the HR email

File: app/Console/Kernel.php [hunk 8]
```diff
@@ -29,6 +30,8 @@ class Kernel extends ConsoleKernel
         FixedBudgetProject::class,
         SendDailyEffortSummaryForProjectsOnGoogleChat::class,
         RemindProjectMembersToUpdateEffortOnGoogleChat::class,
+        LifecyCleOfMail::class,
```
Comment (P4NK4J) [inline app/Console/Kernel.php:33]: Did you even test it on your local ? The classname is different here ?

File: app/Mail/NotifyExceededLifeCycleEmail.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace App\Mail;
+
+use Illuminate\Bus\Queueable;
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Mail\Mailable;
+use Illuminate\Queue\SerializesModels;
+
+class NotifyExpiredLifeCycleEmail extends Mailable implements ShouldQueue
```
Comment (P4NK4J) [inline app/Mail/NotifyExceededLifeCycleEmail.php:10]: What's the meaning of this classname ? Does it sound relevant to you ? How about `ApplicationLifeCycleNotifcationEmail`

**Review by P4NK4J — CHANGES_REQUESTED** [review]
Left some comments, please check

File: Modules/HR/Emails/LifeCycleOfMail.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    public function handle()
+    {
+        $dates = Application::whereIn('status', ['new', 'in_progress'])->pluck('created_at');
+        $expiredApplicationTotalNumber = 0;
+        foreach ($dates as $date) {
+            $difference_days = $date->diffInDays(now());
+            if ($difference_days > config('hr.time-period.outdated')) {
+                $expiredApplicationTotalNumber += 1;
+            }
+        }
```
Comment (P4NK4J) [inline Modules/HR/Emails/LifeCycleOfMail.php:50]: Just see if we can combine this logic in the application query itself, the query should fetch all the applications, whose created date and current date difference is greater than 15 days.

File: app/Console/Kernel.php [hunk 11]
```diff
@@ -29,6 +30,7 @@ class Kernel extends ConsoleKernel
         FixedBudgetProject::class,
         SendDailyEffortSummaryForProjectsOnGoogleChat::class,
         RemindProjectMembersToUpdateEffortOnGoogleChat::class,
+        LifeCycleOfMail::class,
```
Comment (P4NK4J) [inline app/Console/Kernel.php:33]: Let's rename it to `ApplicationLifeCycleEmailTrigger`

File: app/Console/Kernel.php [hunk 12]
```diff
@@ -54,6 +56,7 @@ protected function schedule(Schedule $schedule)
         $schedule->command('project:remind-to-update-effort')->weekdays()->at('19:00');
         $schedule->command('project:send-daily-effort-summary-google-chat')->weekdays()->at('22:30');
         $schedule->command('project:zero-expected-hours-in-project')->weekly()->tuesdays()->at('11:00');
+        $schedule->command('application:lifecycle')->everyday()->at('08:00');
```
Comment (P4NK4J) [inline app/Console/Kernel.php:59]: Let's rename this to `hr:application-lifecycle-email`

**Review by P4NK4J — CHANGES_REQUESTED** [review]
Left some comments, please take a look. Also, add the variable in ENV sample file, so that it can be added in production ENV as well

**Pull request closed without merging**